docs/website/contents/for-developers/NodeTasks.md

```diff
@@ -0,0 +1,305 @@
+   The server can request transactions from the client.
+
+Generally, blocks flow from upstream to downstream peers, while transactions flow from downstream to upstream peers (in order to allow nodes without any downstream peers to still submit transactions to the network).
+
+#### Diffusion of headers and blocks
+
+```mermaid
+graph LR
+  subgraph PeerA[Upstream Peer A]
+    PeerAChainSync[ChainSync]
+    PeerABlockFetch[BlockFetch]
+  end
+
+  subgraph PeerB[Upstream Peer B]
+    PeerBChainSync[ChainSync]
+    PeerBBlockFetch[BlockFetch]
+  end
+
+  subgraph PeerC[Downstream Peer C]
+    PeerCChainSync[ChainSync]
+    PeerCBlockFetch[BlockFetch]
+  end
+
+  subgraph PeerD[Downstream Peer D]
+    PeerDChainSync[ChainSync]
+    PeerDBlockFetch[BlockFetch]
+  end
+
+  BlockFetchLogic
+
+  PeerAChainSync -- headers --> BlockFetchLogic
+  BlockFetchLogic -- instructs --> PeerABlockFetch
+  PeerBChainSync -- headers --> BlockFetchLogic
+  BlockFetchLogic -- instructs --> PeerBBlockFetch
+
+  ChainSelectionAndStorageLayer[ChainSelection and storage layer]
+  PeerABlockFetch -- blocks --> ChainSelectionAndStorageLayer
+  PeerBBlockFetch -- blocks --> ChainSelectionAndStorageLayer
+
+  ChainSelectionAndStorageLayer -- headers --> PeerCChainSync
+  ChainSelectionAndStorageLayer -- blocks --> PeerCBlockFetch
+  ChainSelectionAndStorageLayer -- headers --> PeerDChainSync
+  ChainSelectionAndStorageLayer -- blocks --> PeerDBlockFetch
+```
+
+Here, BlockFetchLogic is a centralized component that decides whether and in which order to download certain blocks based on the candidate header chains.
+
+#### Diffusion of transactions
+
+```mermaid
+graph LR
+  subgraph PeerA[Upstream Peer A]
+    PeerATxSubmission[TxSubmission]
+  end
+
+  subgraph PeerB[Upstream Peer B]
+    PeerBTxSubmission[TxSubmission]
+  end
+
+  subgraph PeerC[Downstream Peer C]
+    PeerCTxSubmission[TxSubmission]
+  end
+
+  subgraph PeerD[Downstream Peer D]
+    PeerDTxSubmission[TxSubmission]
+  end
+
+  Mempool
+
+  Mempool --> PeerATxSubmission
+  Mempool --> PeerBTxSubmission
+  PeerCTxSubmission --> Mempool
+  PeerDTxSubmission --> Mempool
+```
+
+## Relaying a new block
+
+As a concrete example, we will consider how the node relays a new block B from its upstream peers to its downstream peers.
+We assume that no other blocks arrive in the meantime, and that B extends the node's selection.
+
+We also provide rough numbers on the time budget of the steps for illustration.
+
+ 1. In the ChainSync mini protocol thread for peer X, we receive a `MsgRollForward` with the header of B.
+    We validate the header and add it to the candidate header chain for peer X.
+
+    Time budget: average ~1ms, upper bound ~2ms.
+
+ 2. The BlockFetchLogic wakes up due to a change to a candidate chain, and notices the new header.
+    As the header extends our selection, the BlockFetchLogic tells the BlockFetch client for peer X to download the full block B.
+
+    Time budget: essentially instant.
```

Comment:
We saw this be quite bad during your initial CSJ measurements. But that was just a pathological case, right?

Reply:
Yes, in the normal caught-up case, candidate fragments are very small (length 0 or length 1), so it doesn't suffer from that :+1: